### ISSUE TITLE
Please, re-introduce install.py support on Plone 5

### ISSUE THREAD

**Issue by keul** [plone/Products.CMFPlone#443]
Plone 5 will be the most problematic Plone release for add-ons (probably more complex than Plone 2.0 -> 2.1).

Things to be done to make add-ons compatible to both Plone 4.3 and 5.0 are *a lot* and the simpler (and, let me say, more pythonic) way to handle that unavoidable mess *was* the old good guy `Extensions/install.py` script, where we can pilot which/when generic setup profiles are run.

I don't know the reason why it's has been removed from Plone 5, but I think it is a mistake that is not giving nothing good back.

* The way Plone "choose" a profile to be run when installing (alphabetical order) has always been something opinable.
* we still need `install.py` for uninstall.
* `install.py/install` method it's still supported when running installation from ZMI/portal_quickinstaller so the behavior in not coherent with Plone UI.

This lack force add-ons developers to:

* drop Plone 4.3 support, which will be an epic fail
* support two different branch for Plone 4.3 and Plone 5, which often leads to different releases (bug not fixed in one version)
* invent some super complex way to fix this, like: running stuff inside `import_steps.xml`, which is complex, needs more boilerplate and is not exactly the scope it's intended to be.

This complexity is totally useless IMHO.

**Comment by datakurre** [comment]
@keul Is it really true that install.py support is removed and you are not simply missing ``<five:registerPackage />`` from your add-on?

**Comment by davisagli** [comment]
I think dropping support for Install.py probably happened inadvertently when the add-ons control panel was rewritten as a z3c.form form instead of the old skin template. I agree re-adding it would be useful.

**Comment by datakurre** [comment]
Agreed.

**Comment by vangheem** [comment]
+1 to adding it back

**Comment by datakurre** [comment]
The switch to the old behavior should be quite easy, but the Plone 5 
code seem to have a clear intent to drop support for QuickInstaller and 
go GS only. It's just that only install has been implemented, while 
unininstall and reinstall continue to rely on QI.

Does anyone remember the discussion about this?

@keul Nowadays (also in Plone 4) you can register "volatile" GS steps in 
zcml, which makes it relatively easy to execute custom code during 
profile import (the only issue being the need for checking the existence 
of your "marker" file).

Yet, GS has never had real "uninstall" feature, and "uninstall profile" 
is just a convention. Simple add-ons may still have relied on the magic 
of QI (which does not cover the newst tech like p.a.registry).

I guess, because there's no clear winner solution, it's safest to keep 
backwards compatibility to QI.

Nathan Van Gheem wrote:
> +1 to adding it back

**Comment by vangheem** [comment]
Right. I wouldn't even require <five:registerPackage package="." /> in plone 5.

Another idea is modernizing it. Support Extensions/Install.py for now(but deprecate)--move to install and uninstall methods to be defined at the root of "[z3c.autoinclude.plugin] target = plone" packages.

**Comment by datakurre** [comment]
Modernizing is good idea. Needs a sprint to design, implement and document, though.

In the last year, the trend has been to do as much as possible with GS, but while GS can have python upgradeSteps, it cannot have python install/uninstall steps. We do use custom package-custom import steps (often named "setuphandlers") to do the install part, but that's just abusing then GS framework. And for uninstall, Extensions/Install.py doing custom stuff and running the "uninstall profile" has been the only option (unless the magic in QI is not enough).

I'd guess, it would be good enough to add more modern Python-hooks for install/uninstall and making it easy to import profiles with plone.api there. (Similarly now with QI, when you have Extensions/Install.py:install defined, it's the primary installer and GS-profile must be manually imported in the custom install code.)

**Comment by keul** [comment]
Thanks all for taking part on that thread.
Lot of interesting idea! Volatile GS profiles (how it works?), modernize install with entry points...
Still I hope to see `install.py`, probably a deprecated way but "it just works" and simple to use, reintroduced before Plone 5.0 final.

----

Until then: I found another way, not clean as using `install` function, but enough for my usecase:

* register a "common" profile with GS steps that works on both Plone version ("common" is not the name of the profile itself, see below)
* register *two* "default" profiles but including them conditionally with a

        name="default"
        directory="profiles/plone5"
        zcml:condition="not-have plone5"
   
   and
   
        name="default"
        directory="profiles/plone5"
        zcml:condition="have plone5"
* Both profiles contains a `metadata.xml` that depends on common profile and additional unique GS steps (if any).

**Comment by datakurre** [comment]
@keul I meant "volatile" GS import steps, GS steps registered in ZCML instead of import_steps.xml so that they don't remain in the step registry after the add-on has been removed. Otherwise they are just normal steps, so one need to do the normal "is that my profile?" check in the step.

```
  <gs:importStep
      name="unique.name"
      title="display title"
      description="display description"
      handler=".setuphandlers.myStepHandler">
    <depends name="types"/>
  </gs:importStep>
```

Meanwhile, I did introduce pull to restore the old behavior.
